### ISSUE TITLE
[Bug]: no global menu support in wayland-only sessions (linux)

### ISSUE THREAD

**Issue by slynobody** [electron/electron#34335]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

19

### What operating system are you using?

Other Linux

### Operating System Version

debian 11

### What arch are you using?

x64

### Expected Behavior

the global menu functionality (exporting the menu to the menu field in the menubar) should be used out of the box, if you use a x11 or wayland-flavour of linux (and ui of your choice).

### Actual Behavior

global menu does not show up at all (brave, freetube, signal, heroic) when used in a wayland-only environment.
workaround is falling back to x11 by setting GDK_BACKEND="x11" (while *not* setting --ozone-platform=wayland at all or making it default).
resulting in a linux ui / usage-scenario where a whole bunch of electron-apps are currently not benefiting from wayland features (scaling, UHD-resolution, security) while the whole linux-cosystem has mature wayland-support for nearly 1 year.

**Comment by SamDc73** [comment]
I have same exact problem under Fedora KDE

**Comment by cognitus** [comment]
I can confirm. In kde some has reported too https://bugs.kde.org/show_bug.cgi?id=453600

**Comment by slynobody** [comment]
something new on this?!?

**Comment by Kimiblock** [comment]
I can confirm in Arch KDE.

It seems that the only workaround is to run a electron app in xWayland

**Comment by jothi-prasath** [comment]
still no global menu in Arch wayland gnome

**Comment by Edip1** [comment]
still no global menu in gentoo kde wayland

**Comment by slynobody** [comment]
still no global menu in debian kde wayland.

as wayland is mainstreaming, and this bug reduces usability of linux desktop to a large degree (forcing silly env-workaround solutions in each application-starter of each electron-app, f.e. signal, brave, freetube a.s.o. that potentially reduce security and hinder wayland-adoption) , could you please resolve this? Or at least write a few lines on what could be the reason or what other developers of what other platforms should be contacted (ozone?)? 

please give at least an easy to use workaround, f.e. a simple to use env-variable for all electron-apps?

**Comment by Kimiblock** [comment]
Bump


Sent from my Apple Watch

**Comment by sr-tream** [comment]
bump

**Comment by jothi-prasath** [comment]
bump

**Comment by JohnRTitor** [comment]
I can confirm this on Kubuntu 23.10 as well.
VS code which uses Electron also suffers from this bug. If I pass  `--ozone-platform-hint=auto --ozone-platform=wayland` to `code`, it uses a Wayland session and global menu disappears.

**Comment by kzlouge** [comment]
> I can confirm this on Kubuntu 23.10 as well. 我也可以在 Kubuntu 23.10 上确认这一点。 VS code which uses Electron also suffers from this bug. If I pass `--ozone-platform-hint=auto --ozone-platform=wayland` to `code`, it uses a Wayland session and global menu disappears.使用 Electron 的 VS code 也存在这个错误。如果我传递，它会使用 Wayland 会话，并且全局菜单消失。

The same on Archlinu, Plasma 6.0.3

**Comment by Vilez0** [comment]
bump

**Comment by sr-tream** [comment]
> bump

It's a GTK-side issue. GTK do not support global menu in Wayland session

**Comment by JohnRTitor** [comment]
So perhaps this one should be closed and moved to GTK upstream?

**Comment by Kimiblock** [comment]
Probably have to wait [dbus annotation](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52)

**Comment by jothi-prasath** [comment]
bump

**Comment by ahmed658** [comment]
bump

**Comment by afifvdin** [comment]
bump

**Comment by JohnRTitor** [comment]
> Probably have to wait [dbus annotation](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52)

Then perhaps that issue should be bumped instead

**Comment by zstg** [comment]
bump

**Comment by J4ckTh3R1pper** [comment]
bump

**Comment by cognitus** [comment]
Maybe, is better use the same approach used in firefox, https://bugzilla.mozilla.org/show_bug.cgi?id=1956707

> The appmenu protocol is a better alternative. Although it is a KDE protocol, it is the only such a protocol at the moment. Qt also uses it to announce its dbus menus.

**Comment by ghost** [comment]
I don't know if it's been noticed, but for me on KDE and on wayland, Google Chrome now shows its options on the KDE Global Menu.

**Comment by sr-tream** [comment]
@stereomato May be Google Chrome works via XWayland for you? Or you use workaround like this:

https://bugs.kde.org/show_bug.cgi?id=424485#c65

When menu passed via X11 api, but window rendered in native Wayland

**Comment by Vilez0** [comment]
> [@stereomato](https://github.com/stereomato) May be Google Chrome works via XWayland for you? Or you use workaround like this:
> 
> https://bugs.kde.org/show_bug.cgi?id=424485#c65
> 
> When menu passed via X11 api, but window rendered in native Wayland

I'm using Kde plasma and i can see the global menu on brave too.
My browser is working via wayland (I'm sure because xprop cannot capture it)

<img width="543" height="36" alt="Image" src="https://github.com/user-attachments/assets/d7b6a0d1-6e97-474a-87e0-487f020eaa3f" />

**Comment by ghost** [comment]
> [@stereomato](https://github.com/stereomato) May be Google Chrome works via XWayland for you? Or you use workaround like this:
> 
> https://bugs.kde.org/show_bug.cgi?id=424485#c65
> 
> When menu passed via X11 api, but window rendered in native Wayland

I don't use any workaround that I know of, unless KDE/Chrome does it on its own.

**Comment by nanowuu** [comment]
I believe Chrome/Chromium uses the KDE appmenu protocol too, since it is the only way to do it on Wayland at the moment. Dbus-annotation may well take years to happen, so the best approach right now is probably to use KDE's protocol.

**Comment by Davidaprilio** [comment]
in vscode and github desktop there seems to be no problem with the global menu, both use electron
<img width="666" height="192" alt="Image" src="https://github.com/user-attachments/assets/1dd1cdf0-9247-4e0b-a40a-911b6c5d806b" />

<img width="681" height="747" alt="Image" src="https://github.com/user-attachments/assets/bfba19f3-e555-446b-831b-15d45bdad0f8" />

**Comment by Vilez0** [comment]
> in vscode and github desktop there seems to be no problem with the global menu, both use electron <img alt="Image" width="666" height="192" src="https://private-user-images.githubusercontent.com/63033587/498673570-1dd1cdf0-9247-4e0b-a40a-911b6c5d806b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk5MDUwNzUsIm5iZiI6MTc1OTkwNDc3NSwicGF0aCI6Ii82MzAzMzU4Ny80OTg2NzM1NzAtMWRkMWNkZjAtOTI0Ny00ZTBiLWE0MGEtOTExYjZjNWQ4MDZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA4VDA2MjYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3ZTI2MTJkZjc4MmNiYjBiOTU1MjEyYjQwOTBhNDdkMTNhYTEzNmQ2YTg4MzA0YTk0MDA3YjQ0ZTFiMTQwOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.j33nvm9gSXrtDYufuOUcThihZK6J9WOeGO5_7FOpYSg">
> 
> <img alt="Image" width="681" height="747" src="https://private-user-images.githubusercontent.com/63033587/498673894-bfba19f3-e555-446b-831b-15d45bdad0f8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk5MDUwNzUsIm5iZiI6MTc1OTkwNDc3NSwicGF0aCI6Ii82MzAzMzU4Ny80OTg2NzM4OTQtYmZiYTE5ZjMtZTU1NS00NDZiLTgzMWItMTVkNDViZGFkMGY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA4VDA2MjYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhMzI4MWM4OGQyMjQ5ZTY5MzUwYWZkMzliZGI5YmZiMjM2ODgzZTFlOWExYTRiM2U2M2FlNzYyMGNhZmI0NjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Y0Wn_UzFq5BgjZpq6qomMfvz5GkD0obSqtcLbI6fTkM">

Are you sure that they run under wayland and not xwayland?

**Comment by FrostBird347** [comment]
> Are you sure that they run under wayland and not xwayland?

I assume Davidaprilio was using shiftkey's github desktop fork, which has been abandoned for nearly a year and uses an old build of electron that defaults to X11.
As for vscode it does not have a global menu either on my system. Forcing it to use X11 with the command line arguments `--enable-features=UseOzonePlatform --ozone-platform=x11` does however make it appear, so it's possible that Davidaprilio unknowingly had those arguments in their app shortcut or were using an old version of vscode which had that outdated electron build.

**Comment by ahydronous** [comment]
~~No, some applications are still just broken. I am running the latest KDE with a very recent VSCode, and I get no global menu.~~

I am a dolt and misread the thread. Welp, let's hope either appmenu or dbus annotation comes in soon.

**Comment by maher-elattar** [comment]
bump

**Comment by s-celles** [comment]
## Status check (2026-04) — path is clearer than it looks

> Disclosure: the file-paths and plumbing below were scoped by my AI
> agent (Claude Opus 4.7) via targeted reads of electron/electron and
> chromium/chromium. I've cross-checked the key pointers (the
> `org_kde_kwin_appmenu` files, `dbus_appmenu.cc` in Chrome, the
> existing `global_menu_bar_x11.cc` in Electron) but line-level numbers
> and exact signatures should be re-verified against current `main`
> before any code is written.

Chromium upstream actually finished the hard part in 2025. The pieces
Electron needs already exist:

**Chromium side (already there, no patches needed)**
- `ui/ozone/platform/wayland/host/org_kde_kwin_appmenu.{h,cc}` — binds
  the KDE `org_kde_kwin_appmenu_manager` Wayland protocol
- `ui/platform_window/extensions/wayland_extension.h` — exposes
  `WaylandToplevelExtension::SetAppmenu(service_name, object_path)` /
  `UnsetAppmenu()`
- `ui/ozone/platform/wayland/host/ozone_platform_wayland.cc` sets
  `supports_global_application_menus = (org_kde_kwin_appmenu_manager()
  != nullptr)`
- `components/dbus/menu` — cross-backend menu serializer
- `chrome/browser/ui/views/frame/dbus_appmenu.{cc,h}` +
  `dbus_appmenu_registrar.{cc,h}` — working reference impl. Chrome for
  Linux uses it today; its registrar branches X11 (`RegisterWindow`
  DBus call on xid) vs Wayland (`SetAppmenu` on the toplevel) with no
  new wayland protocol work.

**Electron side (what needs doing)**
- `shell/browser/ui/views/global_menu_bar_x11.{h,cc}` (~420 lines)
  still `dlopen`s `libdbusmenu-glib` and is hardcoded to an xid-keyed
  DBus object path. Needs to be renamed (`GlobalMenuBarLinux` or
  similar), reworked to take a `ui::PlatformWindow*` instead of
  `x11::Window`, and to use `components/dbus/menu` for serialization.
- `shell/browser/ui/views/global_menu_bar_registrar_x11.{h,cc}` (~190
  lines) — registrar layer needs the Wayland branch mirroring
  `DbusAppmenuRegistrar::RegisterMenu`: on X11 keep the existing
  `com.canonical.AppMenu.Registrar.RegisterWindow(xid, path)`, on
  Wayland call `WaylandToplevelExtension::SetAppmenu(bus_name, path)`.
- `shell/browser/native_window_views.{cc,h}` — ~10-line glue so the
  unique_ptr isn't typed to the concrete X11 class and instantiation
  is guarded by `supports_global_application_menus` (which it already
  is).

**Not Electron's problem**
- GNOME Wayland has no equivalent protocol — `wayland-protocols!52` is
  stalled. GNOME users would keep the regression regardless.

**Rough size**
- ~500 lines of C++ rewritten + ~10 lines glue + build file updates.
- Zero Chromium patches.
- Estimate: ~2 weeks for someone familiar with Chromium's views stack,
  3-5 weeks for someone new to the Electron codebase. No new Wayland
  protocol research needed.

**Reality check**
- Sharing this in case it's useful to anyone considering the work. I'm
  **not** volunteering to do the port myself, and no WIP/draft PR
  exists at the time of writing (searched open + closed for `wayland
  menu`, `dbusmenu`, `appmenu`, `kwin appmenu`, `SetAppmenu` — zero
  hits).

Reproduced on VS Code 1.116.0 / Electron current / Debian trixie / KDE
Plasma 6.3.6 Wayland — same symptom as original report.

**Comment by mikilezen** [comment]
htps://github.com/megents/megent

**Comment by acomagu** [comment]
I can still reproduce this with Electron 42.1.0 on KDE Plasma Wayland.

**TL;DR:**  [@s-celles 's comment](https://github.com/electron/electron/issues/34335#issuecomment-4275914956) seems to be correct.

In my case the app does export a DBus menu, but it appears to be associated through the X11 AppMenu registrar path rather than the KDE Wayland appmenu protocol.

Observed:

```text
busctl --user tree :1.97
└─ /com/canonical/menu/1

busctl --user introspect :1.97 /com/canonical/menu/1
com.canonical.dbusmenu interface is present

busctl --user call com.canonical.AppMenu.Registrar \
  /com/canonical/AppMenu/Registrar \
  com.canonical.AppMenu.Registrar GetMenuForWindow u 1

=> ":1.97" "/com/canonical/menu/1"
```

The menu layout is non-empty:

```text
GetLayout => Root -> File / Edit / View / Help
```

But KDE logs:

```text
kded6: Error changing property "_KDE_NET_WM_APPMENU_SERVICE_NAME" on window 1 error code: 3
kded6: Error changing property "_KDE_NET_WM_APPMENU_OBJECT_PATH" on window 1 error code: 3
```

So the menu export itself seems to work, but on native Wayland the menu is not linked to the `wl_surface` via KDE’s `org_kde_kwin_appmenu` protocol. It looks like Electron is still using the X11 registrar/window-id path for this.

**Comment by AverageHelper** [comment]
not stale.

**Comment by marcel-wiechmann** [comment]
Issue still exists

**Comment by leessmin** [comment]
Prevent from being classified as outdated. As of the time of my statement, the problem has not been resolved.

**Comment by tarik02** [comment]
https://github.com/tarik02/electron-builds/blob/master/patches/43/wayland-appmenu.patch

I got pretty small vibe-coded patch that works for me for months. May be useful for someone, maybe at some moment will rewrite it properly manually and file as PR.

**Comment by minortex** [comment]
> https://github.com/tarik02/electron-builds/blob/master/patches/43/wayland-appmenu.patch
> 
> I got pretty small vibe-coded patch that works for me for months. May be useful for someone, maybe at some moment will rewrite it properly manually and file as PR.

@tarik02 
Thanks for sharing! Are there any prebuilt Electron binaries with this patch applied? Building Electron from source is a bit heavy for casual users, so a prebuilt would make it much easier to give this a try.

**Comment by tarik02** [comment]
> > https://github.com/tarik02/electron-builds/blob/master/patches/43/wayland-appmenu.patch
> > I got pretty small vibe-coded patch that works for me for months. May be useful for someone, maybe at some moment will rewrite it properly manually and file as PR.
> 
> [@tarik02](https://github.com/tarik02) Thanks for sharing! Are there any prebuilt Electron binaries with this patch applied? Building Electron from source is a bit heavy for casual users, so a prebuilt would make it much easier to give this a try.

I got some prebuilts (they're not actively maintained). Random binaries on internet, so it's up you whether to trust/not trust them.

http://electron-builds.s3.tarik02.me/electron/43/latest.json

**Comment by minortex** [comment]
> I got some prebuilts (they're not actively maintained). Random binaries on internet, so it's up you whether to trust/not trust them.
> 
> https://garage.tarik02.me/api/browse/electron-builds/electron/43/latest.json

That URL gives me a 401 Unauthorized. Does it need auth, or is it not public anymore?

**Comment by tarik02** [comment]
> > I got some prebuilts (they're not actively maintained). Random binaries on internet, so it's up you whether to trust/not trust them.
> > https://garage.tarik02.me/api/browse/electron-builds/electron/43/latest.json
> 
> That URL gives me a 401 Unauthorized. Does it need auth, or is it not public anymore?

sorry I gave wrong link

here is correct http://electron-builds.s3.tarik02.me/electron/43/latest.json
